### DISCUSSION TITLE
Background portal questions

### DISCUSSION THREAD

**Discussion by azymohliad** [flatpak/xdg-desktop-portal#1188] [Questions]
Hi! I'm implementing the background mode and autostart option for my app, and I have a few questions about the [Background](https://flatpak.github.io/xdg-desktop-portal/docs/#gdbus-org.freedesktop.portal.Background) portal:
1. I guess this might be up to the backend, but usually is it supposed to have a confirmation dialog? I don't have it on GNOME 45 and the permission is granted immediately after the request, without any input from the user. Is it going to change later? Is it different on other portal backends? Wondering how important it is to send the correct window handle in the request.
2. If this portal isn't supposed to have a confirmation dialog, are there any other possible reasons that requested permissions might not be granted? I'm considering how diligently I should handle the unexpected response.
3. If the background permission is already granted, and the app sends an additional request for autostart permission, is it possible that the background permission in the response is revoked? E.g. if there's another confirmation dialog, and the user rejects. Wondering if I should handle the `background` response field for autostart-only requests (from the app-logic PoV).
4. Just a general curiosity: the API is designed in such a way that it's impossible to request autostart without requesting background permission. Why is that? I guess it makes little sense for the app to autostart and not be allowed to run in the background, but I can imagine use some cases for that. A minor inconvenience of having both in a single request for the app development, is that if you want to make sure the background permission is granted but without affecting the autostart state, you have to additionally send the correct current autostart setting too. But that's not too big of an inconvenience of course

**Comment by Mikenux** [comment]
Hello!

Backends can decide to grant permission. In GNOME it was decided to grant background permission by default, whereas before this was asked to the user; it's not going to change. I don't know what the other backends do.

> are there any other possible reasons that requested permissions might not be granted?

Personally, I've never read anywhere that a backend automatically denies to grant permission. However, a desktop environment may not have a backend implemented (so no permissions can be granted).

> 3\. If the background permission is already granted, and the app sends an additional request for autostart permission, is it possible that the background permission in the response is revoked?

Given https://github.com/flatpak/xdg-desktop-portal/blob/main/data/org.freedesktop.portal.Background.xml, it seems the request is for both at the same time. For example, I wasn't asked in GNOME for autostart with the Blanket app (I tested).

I have no other answer to your other questions because I lack the knowledge for.

**Reply by azymohliad** [reply]
Thanks a lot for the answers!

Regarding the 3rd question, yes, GNOME doesn't ask me for confirmation either whether I pass `autostart` as `true` or `false` in the requests. Tho I'm still wondering if it's possible that after a successful background-only request (`autostart = false`), another background + autostart request may respond with `background = false` (e.g. if there's a confirmation dialog on another backend).

It's just that I have two separate switches for background and autostart in my app settings, and two separate protal response handlers for each. And I'm wondering if I should worry about background switch state in my autostart response handler. I guess that's up to backend implementation, and I wouldn't worry about it for now

**Reply by Mikenux** [reply]
Still on GNOME, if I disable the background permission for Blanket, then I have a dialog telling me to enable the background permission when I want to enable autostart from Blanket.
